prov/shm/src/smr_progress.c

```diff
@@ -463,12 +463,19 @@ static int smr_progress_ipc(struct smr_cmd *cmd, enum fi_hmem_iface iface,
 						  device, iov, iov_count, 0, ptr,
 						  cmd->msg.hdr.size);
 	}
-	if (!ret)
-		*total_len = cmd->msg.hdr.size;
 
 	if (cmd->msg.data.ipc_info.iface == FI_HMEM_ZE)
 		close(ipc_fd);
+
+	/* Do not go directly to out if this results in an error. The
+	 * error due to truncation should take precedence if applicable.
+	 */
 	ret = ofi_hmem_close_handle(cmd->msg.data.ipc_info.iface, base);
+
+	if (*total_len != cmd->msg.hdr.size) {
+		ret = (*total_len < 0) ? ((int) *total_len) : -FI_ETRUNC;
+	}
+
```

Comment:
When ofi_copy_to_hmem_iov returns error, it's sufficient to return that error code. Do we really want *total_len be set to the error code?  0 is probably more appropriate.

Reply:
I was under the impression that the value of `total_len` is not considered valid if the function returns an error. If that is incorrect, or `total_len` should hold a different value I can make that change.

Reply:
yes, the error code takes precedence - total_len will be ignored if err is set